tripal/t4d8
# Support for Materialized Views.

Issue
===
Issue #212 

This PR should not be merged until after PR #244 but is ready for review.  However, it may be best to review after PR #244 is merged because all of the code from that PR is also shown here in the "Files Changed" tab.

Changes
======
The following changes were made in this PR.
1.  **A new ChadoCustomTable Service** has been added.  Previously the older custom table API functions were updated but when it came time to add in the materialized view it just seemed clunky because both the Mviews and the custom table share a lot of the same functionality.  This class manages interactions with the `tripal_custom_tables` table.  This class is chado schema aware so custom tables can be created in any Chado schema.
2. **A new ChadoMview Service** has been added. This class extends the ChadoCustomTable class to provide additional support for materialized views.   This class is chado schema aware so mviews can be created in any Chado schema.
3. **Database changes** were made to the `tripal_custom_tables` and `tripal_mviews` tables. Here's what's different:
    1. I removed the `mview_id` from the `tripal_custom_tables` because it seems to me the relationship should go the other way since an mview is a custom table but  custom table is not always an mview.
    2. I added a `hidden` column to the `tripal_custom_tables` because some custom tables are used internally and should not be changed or deleted by an end-user. For example, all of the temp tables used by the GFF importer as well as some of the mviews (e.g. db2cv_mview).  Any tables where hidden is set to `1` will not be visible to the end-user but can still be managed programmatically.
    3. I add a `table_id` column to the `tripal_mviews` table to create the foreign key.
    4. I removed old and unused historical columns from the `tripal_mviews` table. These include:  `mv_specs`, `special_index`, `indexed` and `modulename`
4. **The old API is now deprecated** both for custom tables and materialized views.  The functions are still there and have been updated to use the new classes but they do need some functional testing.   In a previous PR someone had added a new variable to these API functions to pass in the Chado schema.  However, I removed that to restore the functions to their Tripal v3 form and left a warning in the function inline documentation to indicate that the functions only work on the default Chado instance. This is actually consistent with the Tripal v3 way. And by not having that functionality it encourages use of the new classes.
5. **New Forms** now exist for creating, editing, deleting and populating materialized views.
6. **Updates to Forms** for creating, editing and deleting were made to use the new classes.
7. **A new Materialized Views View** was created so that admins can see the current list of materialized views.
8. **Updates to the Custom Tables View** were made so that hidden tables don't appear and materialized views will not either.  The user must edit materialized views now using the MViews view.
9. **New User Permissions** for managing custom tables and materialized views. This way site admins can be more precise in who they give permission to create new tables.
10. **Dropping a Chado schema** will now remove the entries in both the `tripal_custom_tables` and `tripal_mviews` tables. This way, if a schema is dropped those tables don't still show up in the list.
11. **Preparing Chado** now supports creation of mviews and population of the `db2cv_mview` and the `cv_root_mview`

What's Not Included
=============
I just didn't have time to add the following and I didn't want to delay getting this functionality into Tripal. 

1. Developer Documentation for how to programatically create custom tables and materialized views
2. Design documentation for the ChadoCustomTable and ChadoMview classe.s
3. Functional testing.
4. Code to migrate Tripal v3 `tripal_custom_tables` and `tripal_mviews` to the new structure in Tripal v4. I'm not sure yet how we're going to write the upgrade code.

How to Test
========
1. Pull this branch.
2. Start with a completely new Tripal installation. This will guarantee you have the new `tripal_mviews` table and updated `tripal_custom_tables` table.  
3. Install a new Chado instance and run the prepare. After the prepare check the following:
   1. The output on the command-line should indicate that materialized views were created.
   2. The output on the command-line should indicate that the `db2cv_mview` and `cv_root_mview` were populated.
   3. Go to the Chado schema (via phppgadmin or psql) and make sure that those two mviews are present and that they have data.
   4. Go to Admin >> Tripal >> Data Storage >> Chado >> Custom tables.  This should show an empty table. This is because the custom tables that were created should be hidden.
   5. Go to the Chado schema (via phppgadmin or psql) and make sure the `tripal_xxxx_temp` tables are created.
   6. Go to the public schema (via phppgadmin or psql) and make sure the `tripal_xxxx_temp` tables are present and marked as hidden. This is why they don't appear in the online GUI for managing custom tables.
   7. Go to Admin >> Tripal >> Data Storage >> Chado >> Materialized views.  You should see a few tables listed.  Using steps similar to 5 & 6, verify that the tables exist in Chado and in the `tripal_mviews` table.  
4. Managing  Custom Tables. Go to Admin >> Tripal >> Data Storage >> Chado >> Custom tables
   1.   Create a new table.  it should appear in the list and be present in Chado.
   2.   Edit the table, change the table name but do not check 'recreate'.  It should fail and the error message should be that you can't change the schema if you change the table name.
   3.  Edit the table, change the table name and check the 'recreate'. The table should be renamed both in the GUI and in Chado.
   4. Delete the table. it should be gone from the GUI and from Chado.
   5. Find a new table to edit. Change the schema, and check the 'recreate' box.  The table schema should now be saved as you set and the table in Chado should have the new changes.
6. Managing  Custom Tables. Go to Admin >> Tripal >> Data Storage >> Chado >> Materialized Views
   1.  Repeat the same tests for managing custom tables above.
   2.  Test populating of any mview.  It should submit a Tripal job. Run the job.  Return to the listing of Mviews. It should now show the date last updated and that 0 rows were added (i.e. becasue there is not data in Chado).
7.. Working with another Chado  
    1.  Create new Chado instance and populate it.
    2.  Repeat steps 4-5 above to make sure that the changes occur in the second Chado instance rather than the first.
    3.  Drop the second Chado schema.  Return to custom tables listing and the materialized view listing and all tables in the dropped schema should be gone.
8. Do any other crazy thing you want.
